tune-operating-system.md

```diff
@@ -0,0 +1,120 @@
+
+### BCC/bpftrace
+
+Starting from CentOS 7.6, the Linux kernel has supported Berkeley Packet Filter (BPF). Therefore, you can choose proper tools to conduct an in-depth analysis based on the results in [In 60 seconds](#in-60-seconds). Compared with perf/ftrace, BPF provides programmability and smaller performance overhead. Compared with kprobe, BPF provides higher security and is more suitable for the production environments. For detailed usage of the BCC toolkit, see [BPF Compiler Collection (BCC)](https://github.com/iovisor/bcc/blob/master/README.md).
+
+## Performance tuning
+
+This section introduces performance tuning based on the classified kernel subsystems.
+
+### CPU -- frequency scaling
+
+cpufreq is a module that dynamically adjusts the CPU frequency and supports five modes. To ensure service performance, select the performance mode and fix the CPU frequency at the highest supported operating frequency without dynamic adjustment. The command for this operation is `cpupower frequency-set --governor performance`.
+
+### CPU -- interrupt affinity
+
+- Automatic balance can be implemented through the `irqbalance` service.
+- Manual balance:
+    - Identify the devices that need to balance interrupts. Starting from CentOS 7.5, the system automatically configures the best interrupt affinity for certain devices and their drivers, such as devices that use the `be2iscsi` driver and NVMe settings. You can no longer manually configure interrupt affinity for such devices.
+    - For other devices, check the chip manual to see whether these devices support distributing interrupts.
+        - If they do not, all interrupts of these devices are routed to the same CPU and cannot be modified.
+        - If they do, calculate the `smp_affinity` mask and set the corresponding configuration file. For details, see [kernel document](https://www.kernel.org/doc/Documentation/IRQ-affinity.txt).
+
+### NUMA CPU binding
+
+To avoid accessing memory across Non-Uniform Memory Access (NUMA) nodes as much as possible, you can bind a thread/process to certain CPU cores by setting the CPU affinity of the thread. For ordinary programs, you can use the `numactl` command for the CPU binding. For detailed usage, see the Linux manual pages. For network interface card (NIC) interrupts, see [tune network](#tune-network).
+
+### Memory -- transparent huge page (THP)
+
+It is **NOT** recommended to use THP for database applications, because databases often have sparse rather than continuous memory access patterns. If high-level memory fragmentation is serious, a higher latency will occur when THP pages are allocated. If the direct compaction is enabled for THP, the CPU usage will surge. Therefore, it is recommended to disable the direct compaction for THP.
```

Comment:
The Chinese version of the last sentence is confusing. Please confirm it.

Reply:
As confirmed with @ethercflow, `disable THP` is correct.
```suggestion
It is **NOT** recommended to use THP for database applications, because databases often have sparse rather than continuous memory access patterns. If high-level memory fragmentation is serious, a higher latency will occur when THP pages are allocated. If the direct compaction is enabled for THP, the CPU usage will surge. Therefore, it is recommended to disable THP.
```